File: src/main/scala/com/github/phisgr/gatling/grpc/stream/TimestampExtractor.scala
package com.github.phisgr.gatling.grpc.stream

import com.github.phisgr.gatling.grpc.util.toProtoString
import com.typesafe.scalalogging.Logger
import io.gatling.commons.stats.Status
import io.gatling.core.session.Session
import io.gatling.core.stats.StatsEngine

trait TimestampExtractor[-Res] extends EventExtractor[Res] {
  /**
   * @param session         The session of the streaming branch.
   * @param message         The message received from the streaming call
   * @param streamStartTime The start time of this stream
   * @return The "start time" of this message,
   *         or [[TimestampExtractor.IgnoreMessage]] if this message should not be logged.
   */
  def extractTimestamp(session: Session, message: Res, streamStartTime: Long): Long

  final override def writeEvents(
    session: Session,
    streamStartTime: Long,
    requestName: String,
    message: Res,
    receiveTime: Long,
    statsEngine: StatsEngine,
    logger: Logger,
    status: Status,
    errorMessage: Option[String]
  ): Unit = {
    val extractedTime = extractTimestamp(session, message, streamStartTime)

    if (extractedTime == TimestampExtractor.IgnoreMessage) {
      logger.trace(s"Ignored message\n${toProtoString(message)}")
    } else {
      statsEngine.logResponse(
        session.scenario,
        session.groups,
        requestName = requestName,
        startTimestamp = extractedTime,
        endTimestamp = receiveTime,
        status = status,
        responseCode = None,
        message = errorMessage
      )
    }

  }
}

object TimestampExtractor {
  /** Return this sentinel value and the message will not be logged */
  final val IgnoreMessage = Long.MinValue

  final val Ignore: TimestampExtractor[Any] = (_, _, _) => IgnoreMessage

  /**
   * Java/Kotlin API.
   * Java and Kotlin compilers don't know that
   * [[TimestampExtractor]] is contravariant.
   */
  def ignore[T]: TimestampExtractor[T] = Ignore
}

File: gatling-grpc-kt/src/main/kotlin/com/github/phisgr/gatling/kt/grpc/stream/TimestampExtractor.kt
package com.github.phisgr.gatling.kt.grpc.stream

import io.gatling.javaapi.core.Session
import com.github.phisgr.gatling.grpc.stream.TimestampExtractor as TimestampExtractorS
import io.gatling.core.session.Session as SessionS

@FunctionalInterface
fun interface TimestampExtractor<in Res> : TimestampExtractorS<@UnsafeVariance Res> {
    companion object {
        @JvmField
        val IGNORE: TimestampExtractor<Any?> = TimestampExtractor { _, _, _ -> IGNORE_MESSAGE }

        /**
         * See [TimestampExtractorS.IgnoreMessage]
         */
        @JvmField
        val IGNORE_MESSAGE = TimestampExtractorS.IgnoreMessage()
    }

    /**
     * See [TimestampExtractorS.extractTimestamp]
     */
    fun extractTimestamp(session: Session, message: Res, streamStartTime: Long): Long

    override fun extractTimestamp(session: SessionS, message: Res, streamStartTime: Long): Long =
        extractTimestamp(Session(session), message, streamStartTime)
}
